File: slang-checks/src/main/java/org/sonarsource/slang/checks/complexity/CognitiveComplexity.java
/*
 * SonarSource SLang
 * Copyright (C) 2018-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonarsource.slang.checks.complexity;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.Iterator;
import java.util.List;
import java.util.Set;
import javax.annotation.Nullable;
import org.sonarsource.slang.api.BinaryExpressionTree;
import org.sonarsource.slang.api.CatchTree;
import org.sonarsource.slang.api.ClassDeclarationTree;
import org.sonarsource.slang.api.FunctionDeclarationTree;
import org.sonarsource.slang.api.IfTree;
import org.sonarsource.slang.api.LoopTree;
import org.sonarsource.slang.api.MatchTree;
import org.sonarsource.slang.api.Token;
import org.sonarsource.slang.api.Tree;
import org.sonarsource.slang.checks.utils.ExpressionUtils;
import org.sonarsource.slang.impl.JumpTreeImpl;
import org.sonarsource.slang.visitors.TreeContext;
import org.sonarsource.slang.visitors.TreeVisitor;

import static org.sonarsource.slang.checks.utils.ExpressionUtils.isLogicalBinaryExpression;

public class CognitiveComplexity {

  private List<Increment> increments = new ArrayList<>();

  public CognitiveComplexity(Tree root) {
    CognitiveComplexityVisitor visitor = new CognitiveComplexityVisitor();
    visitor.scan(new TreeContext(), root);
  }

  public int value() {
    int total = 0;
    for (Increment increment : increments) {
      total += increment.nestingLevel + 1;
    }
    return total;
  }

  public List<Increment> increments() {
    return increments;
  }

  public static class Increment {

    private final Token token;
    private final int nestingLevel;

    private Increment(Token token, int nestingLevel) {
      this.token = token;
      this.nestingLevel = nestingLevel;
    }

    public Token token() {
      return token;
    }

    public int nestingLevel() {
      return nestingLevel;
    }
  }

  private class CognitiveComplexityVisitor extends TreeVisitor<TreeContext> {

    private Set<Token> alreadyConsideredOperators = new HashSet<>();

    private CognitiveComplexityVisitor() {

      // TODO ternary operator
      // TODO "break" or "continue" with label

      register(LoopTree.class, (ctx, tree) -> incrementWithNesting(tree.keyword(), ctx));
      register(MatchTree.class, (ctx, tree) -> incrementWithNesting(tree.keyword(), ctx));
      register(CatchTree.class, (ctx, tree) -> incrementWithNesting(tree.keyword(), ctx));
      register(JumpTreeImpl.class, (ctx, tree) -> {
        if (tree.label() != null) {
          incrementWithoutNesting(tree.keyword());
        }
      });

      register(IfTree.class, (ctx, tree) -> {
        Tree parent = ctx.ancestors().peek();
        boolean isElseIf = parent instanceof IfTree && tree == ((IfTree) parent).elseBranch();
        boolean isTernary = ExpressionUtils.isTernaryOperator(ctx.ancestors(), tree);
        if (!isElseIf || isTernary) {
          incrementWithNesting(tree.ifKeyword(), ctx);
        }
        Token elseKeyword = tree.elseKeyword();
        if (elseKeyword != null && !isTernary) {
          incrementWithoutNesting(elseKeyword);
        }
      });

      register(BinaryExpressionTree.class, (ctx, tree) -> handleBinaryExpressions(tree));
    }

    private void handleBinaryExpressions(BinaryExpressionTree tree) {
      if (!isLogicalBinaryExpression(tree) || alreadyConsideredOperators.contains(tree.operatorToken())) {
        return;
      }

      List<Token> operators = new ArrayList<>();
      flattenOperators(tree, operators);

      Token previous = null;
      for (Token operator : operators) {
        if (previous == null || !previous.text().equals(operator.text())) {
          incrementWithoutNesting(operator);
        }
        previous = operator;
        alreadyConsideredOperators.add(operator);
      }
    }

    // TODO parentheses should probably be skipped
    private void flattenOperators(BinaryExpressionTree tree, List<Token> operators) {
      if (isLogicalBinaryExpression(tree.leftOperand())) {
        flattenOperators((BinaryExpressionTree) tree.leftOperand(), operators);
      }

      operators.add(tree.operatorToken());

      if (isLogicalBinaryExpression(tree.rightOperand())) {
        flattenOperators((BinaryExpressionTree) tree.rightOperand(), operators);
      }
    }

    private void incrementWithNesting(Token token, TreeContext ctx) {
      increment(token, nestingLevel(ctx));
    }

    private void incrementWithoutNesting(Token token) {
      increment(token, 0);
    }

    private void increment(Token token, int nestingLevel) {
      increments.add(new Increment(token, nestingLevel));
    }

    private int nestingLevel(TreeContext ctx) {
      int nestingLevel = 0;
      boolean isInsideFunction = false;
      Iterator<Tree> ancestors = ctx.ancestors().descendingIterator();
      Tree parent = null;
      while (ancestors.hasNext()) {
        Tree t = ancestors.next();
        if (t instanceof FunctionDeclarationTree) {
          if (isInsideFunction || nestingLevel > 0) {
            nestingLevel++;
          }
          isInsideFunction = true;
        } else if ((t instanceof IfTree && !isElseIfBranch(parent, t)) || t instanceof MatchTree || t instanceof LoopTree || t instanceof CatchTree) {
          nestingLevel++;
        } else if (t instanceof ClassDeclarationTree) {
          nestingLevel = 0;
          isInsideFunction = false;
        }
        parent = t;
      }
      return nestingLevel;
    }

    private boolean isElseIfBranch(@Nullable Tree parent, Tree t) {
      return parent instanceof IfTree && ((IfTree) parent).elseBranch() == t;
    }

  }

}

File: sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/CognitiveComplexity.kt
/*
 * SonarSource SLang
 * Copyright (C) 2018-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonarsource.kotlin.checks

import org.jetbrains.kotlin.com.intellij.psi.PsiElement
import org.jetbrains.kotlin.lexer.KtTokens
import org.jetbrains.kotlin.psi.KtBinaryExpression
import org.jetbrains.kotlin.psi.KtBlockExpression
import org.jetbrains.kotlin.psi.KtBreakExpression
import org.jetbrains.kotlin.psi.KtCatchClause
import org.jetbrains.kotlin.psi.KtClass
import org.jetbrains.kotlin.psi.KtContinueExpression
import org.jetbrains.kotlin.psi.KtElement
import org.jetbrains.kotlin.psi.KtExpression
import org.jetbrains.kotlin.psi.KtForExpression
import org.jetbrains.kotlin.psi.KtFunction
import org.jetbrains.kotlin.psi.KtIfExpression
import org.jetbrains.kotlin.psi.KtLoopExpression
import org.jetbrains.kotlin.psi.KtOperationReferenceExpression
import org.jetbrains.kotlin.psi.KtTreeVisitorVoid
import org.jetbrains.kotlin.psi.KtWhenExpression
import org.jetbrains.kotlin.psi.KtWhileExpression

/**
 * Replacement for [org.sonarsource.slang.checks.complexity.CognitiveComplexity]
 */
class CognitiveComplexity(val root: KtElement) {
    private val increments: MutableList<Increment> = ArrayList()

    init {
        val visitor = CognitiveComplexityVisitor()
        root.accept(visitor)
    }

    fun value(): Int {
        var total = 0
        for (increment in increments) {
            total += increment.nestingLevel + 1
        }
        return total
    }

    fun increments(): List<Increment> {
        return increments
    }

    class Increment(val token: PsiElement, val nestingLevel: Int)

    private inner class CognitiveComplexityVisitor : KtTreeVisitorVoid() {
        private val alreadyConsideredOperators: MutableSet<KtOperationReferenceExpression> = HashSet()

        override fun visitExpression(expression: KtExpression) {
            when (expression) {
                is KtForExpression -> incrementWithNesting(expression.forKeyword, expression)
                is KtWhileExpression -> incrementWithNesting(expression.firstChild, expression)
                is KtWhenExpression -> incrementWithNesting(expression.whenKeyword, expression)
                is KtBreakExpression -> expression.labelQualifier?.let { incrementWithoutNesting(expression.firstChild) }
                is KtContinueExpression -> expression.labelQualifier?.let { incrementWithoutNesting(expression.firstChild) }
                is KtIfExpression -> handleIfExpression(expression)
                is KtBinaryExpression -> handleBinaryExpressions(expression)
            }
            super.visitExpression(expression)
        }

        override fun visitCatchSection(catchClause: KtCatchClause) {
            incrementWithNesting(catchClause.firstChild, catchClause)
            super.visitCatchSection(catchClause)
        }

        private fun handleBinaryExpressions(tree: KtBinaryExpression) {
            if (!tree.isLogicalBinaryExpression() || alreadyConsideredOperators.contains(tree.operationReference)) {
                return
            }
            val operators: MutableList<KtOperationReferenceExpression> = ArrayList()
            flattenOperators(tree, operators)
            var previous: KtOperationReferenceExpression? = null
            for (operator in operators) {
                if (previous == null || previous.operationSignTokenType?.value != operator.operationSignTokenType?.value) {
                    incrementWithoutNesting(operator)
                }
                previous = operator
                alreadyConsideredOperators.add(operator)
            }
        }

        private fun flattenOperators(tree: KtBinaryExpression, operators: MutableList<KtOperationReferenceExpression>) {
            if (tree.left.isLogicalBinaryExpression()) {
                flattenOperators(tree.left as KtBinaryExpression, operators)
            }
            operators.add(tree.operationReference)
            if (tree.right.isLogicalBinaryExpression()) {
                flattenOperators(tree.right as KtBinaryExpression, operators)
            }
        }

        private fun incrementWithNesting(token: PsiElement, tree: KtElement) {
            increment(token, nestingLevel(tree))
        }

        private fun incrementWithoutNesting(token: PsiElement) {
            increment(token, 0)
        }

        private fun increment(token: PsiElement, nestingLevel: Int) {
            increments.add(Increment(token, nestingLevel))
        }

        private fun nestingLevel(ktElement: KtElement): Int {
            var nestingLevel = 0
            var parent: PsiElement? = ktElement.parent
            var isInsideFunction = root is KtFunction

            val ancestors = ArrayDeque<PsiElement>()
            while (parent != null && parent != root) {
                ancestors.addFirst(parent)
                parent = parent.parent
            }

           for (element in ancestors) {
               if (element is KtFunction) {
                   if (isInsideFunction) nestingLevel++
                   isInsideFunction = true
               } else if (element is KtIfExpression && !isElseIfBranch(element.parent?.parent, element) ||
                   element is KtWhenExpression || element is KtLoopExpression || element is KtCatchClause) {
                   nestingLevel++
               } else if (element is KtClass) {
                   nestingLevel = 0
                   isInsideFunction = false
               }
           }
            return nestingLevel
        }

        private fun isElseIfBranch(parent: PsiElement?, t: KtElement): Boolean {
            return parent is KtIfExpression && parent.`else` === t
        }

        private fun handleIfExpression(ifExpression: KtIfExpression) {
            val parent = ifExpression.parent
            val preParent = parent?.parent
            val isElseIf = preParent is KtIfExpression && ifExpression === preParent.`else`
            if (!isElseIf) {
                incrementWithNesting(ifExpression.ifKeyword, ifExpression)
            }
            ifExpression.`else`?.let {
                if (it is KtBlockExpression || it is KtIfExpression)
                    incrementWithoutNesting(ifExpression.elseKeyword!!)
            }
        }
    }

    private fun KtExpression?.isLogicalBinaryExpression(): Boolean =
        this is KtBinaryExpression && (operationToken == KtTokens.ANDAND || operationToken == KtTokens.OROR)
}
